specs/core/0_beacon-chain.md

```diff
@@ -1772,135 +1772,291 @@ For each `transfer` in `block.body.transfers`:
+    total_balance = get_previous_total_balance(state)
+    total_attesting_balance = get_attesting_balance(get_previous_epoch_attestations(state))
+    matching_head_attestations = [
+        a for a in get_previous_epoch_attestations(state) if
+        a.data.beacon_block_root == get_block_root(state, a.data.slot)
+    ]
+    matching_head_balance = get_attesting_balance(matching_head_attestations)
+    # Normal case
+    if epochs_since_finality <= 4:
+        for index in active_validator_indices:
+            # Expected FFG source
+            if index in get_attesting_indices(get_previous_epoch_attestations(state)):
+                deltas[index] += get_base_reward(state, index) * total_attesting_balance // total_balance
+                # Inclusion speed bonus
+                deltas[index] += (
+                    get_base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
+                    inclusion_distance(state, index)
+                )
+            else:
+                deltas[index] -= get_base_reward(state, index)
+            # Expected FFG target
+            if index in get_attesting_indices(boundary_attestations):
+                deltas[index] += get_base_reward(state, index) * boundary_attesting_balance // total_balance
+            else:
+                deltas[index] -= get_base_reward(state, index)
+            # Expected head
+            if index in matching_head_attestations:
+                deltas[index] += get_base_reward(state, index) * matching_head_balance // total_balance
+            else:
+                deltas[index] -= get_base_reward(state, index)    
+            # Proposer bonus
+            proposer_index = get_beacon_proposer_index(state, inclusion_slot(state, index))
+            deltas[proposer_index] += base_reward(state, index) // ATTESTATION_INCLUSION_REWARD_QUOTIENT
+    # Quadratic leak case
+    else:
+        for index in active_validator_indices:
+            if index not in get_attesting_indices(get_previous_epoch_attestations(state)):
+                deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)
+            else:
+                # If a validator did attest, apply a small penalty for getting attestations included late
+                deltas[index] += (
+                    base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
+                    inclusion_distance(state, index)
+                )
+                deltas[index] -= base_reward(state, index)
+                    
+            if index not in get_attesting_indices(boundary_attestations):
+                deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)
+            if index not in matching_head_attestations:
+                deltas[index] -= get_base_reward(state, index)
+        for index in range(len(state.validator_registry)):
+            if index not in active_validator_indices and state.validator_registry[index].slashed:
```

Comment:
```suggestion
            if index in active_validator_indices and state.validator_registry[index].slashed:
```

Reply:
That clause is specifically there for inactive validators; active validators get slashed in the section right above.

Reply:
Hmm… but the original condition states that `Any active validator index with validator.slashed is True, loses 2 * inactivity_penalty(state, index, epochs_since_finality) + base_reward(state, index).`.

Also the comment `# Penalize slashed-but-inactive validators as though they were active but offline` seems a little confusing to me or I probably have misunderstood what `active` means in this context.

Reply:
I actually think that's wrong and it should be any inactive and not-withdrawn validator loses that amount. The intention is to ensure that you can't get away from inactivity penalties by self-slashing to instantly log yourself out; you still have to endure them.